include/shell.h

```diff
@@ -20,134 +20,164 @@
 #define DOSBOX_SHELL_H
 
-#include "dosbox.h"
-
-#include <cctype>
-#include <list>
 #include <memory>
+#include <optional>
+#include <stack>
 #include <string>
 
-#ifndef DOSBOX_PROGRAMS_H
+#include "callback.h"
 #include "programs.h"
-#endif
 
-#include "callback.h"
-#include "help_util.h"
+class DOS_Shell;
+
+constexpr auto CMD_MAXLINE = 4096;
 
-#define CMD_MAXLINE 4096
-#define CMD_MAXCMDS 20
-#define CMD_OLDSIZE 4096
 extern callback_number_t call_shellstop;
-class DOS_Shell;
 
-/* first_shell is used to add and delete stuff from the shell env 
+/* first_shell is used to add and delete stuff from the shell env
  * by "external" programs. (config) */
-extern DOS_Shell * first_shell;
+extern DOS_Shell* first_shell;
+
+class ByteReader {
+public:
+	virtual void Reset()               = 0;
+	virtual std::optional<char> Read() = 0;
+
+	ByteReader()                             = default;
+	ByteReader(const ByteReader&)            = delete;
+	ByteReader& operator=(const ByteReader&) = delete;
+	ByteReader(ByteReader&&)                 = delete;
+	ByteReader& operator=(ByteReader&&)      = delete;
+	virtual ~ByteReader()                    = default;
+};
 
+class HostShell {
+public:
+	virtual bool GetEnvStr(const char* entry, std::string& result) const = 0;
+
+	HostShell()                            = default;
+	HostShell(const HostShell&)            = delete;
+	HostShell& operator=(const HostShell&) = delete;
+	HostShell(HostShell&&)                 = delete;
+	HostShell& operator=(HostShell&&)      = delete;
+	virtual ~HostShell()                   = default;
+};
 
 class BatchFile {
 public:
-	BatchFile(DOS_Shell * host,char const* const resolved_name,char const* const entered_name, char const * const cmd_line);
-	BatchFile(const BatchFile&) = delete; // prevent copying
-	BatchFile& operator=(const BatchFile&) = delete; // prevent assignment
-	virtual ~BatchFile();
-	virtual bool ReadLine(char * line);
+	BatchFile(const HostShell& host, std::unique_ptr<ByteReader> input_reader,
+	          std::string_view entered_name, std::string_view cmd_line,
+	          bool echo_on);
+	BatchFile(const BatchFile&)            = delete;
+	BatchFile& operator=(const BatchFile&) = delete;
+	BatchFile(BatchFile&&)                 = delete;
+	BatchFile& operator=(BatchFile&&)      = delete;
+	virtual ~BatchFile()                   = default;
+
+	virtual bool ReadLine(char* line);
 	bool Goto(std::string_view label);
 	void Shift();
-	uint16_t file_handle = 0;
-	uint32_t location = 0;
-	bool echo = false;
-	DOS_Shell *shell = nullptr;
-	std::shared_ptr<BatchFile> prev = {}; // shared with Shell.bf
-	std::unique_ptr<CommandLine> cmd = {};
-	std::string filename{};
+	void SetEcho(bool echo_on);
+	[[nodiscard]] bool Echo() const;
 
 private:
 	[[nodiscard]] std::string ExpandedBatchLine(std::string_view line) const;
 	[[nodiscard]] std::string GetLine();
+
+	const HostShell& shell;
+	CommandLine cmd;
+	std::unique_ptr<ByteReader> reader;
+	bool echo;
```

Comment:
Please use defaults initializers.

Reply:
> All of these members are completely dependent on what the constructor gives them.

OK; perfect. We have plenty of places where that's the case for the same reason.  (maybe mention that in a comment here though).

Usually class implementations (in general) are so sloppy that there are many ways to construct an object. Ideally I would like explicitly invalid defaults which get asserted as non-invalid. That's not needed here because of how well locked down you've made the class, which is great. 

 Can the CMD line and bool be const then?  (I realize the reader cannot because it updates per line read, like you explained before).